src/compiler/ir/README.md

```diff
@@ -0,0 +1,162 @@
+# General structure
+
+The IR has a fairly standard structure, made up of basic blocks. Each basic
+block has a list of statement and a terminator. While the implementation
+occasionally refers to "phi nodes", the IR uses basic block arguments rather
+than phi nodes.
+
+The IR only represents the body of methods. The "shape" of the program, eg. its
+classes, fields, method names and signatures, ... are all stored in the AST
+instead. For each `Method` node in the AST, a corresponding `FnAnalysis` is
+created, which, among others, contains the `MethodIR`.
+
+A `MethodIR` is composed of one or more `FunctionIR`. This distinction is used
+to encode the bodies of `when` clauses that appear within the method. The first
```

Comment:
Lambdas aren't implemented yet. I'm still undecided how they should be represented.

In theory both lambdas and when clauses lower to an anonymous class. However `when` clauses are "easier" to implement because the program never receives a reference to that object, therefore we don't need to give when clauses any type, which made it very easy to implement. Lambdas will need more thinking.

Reply:
Btw I'm thinking we should rename `FunctionIR` since I find the naming confusing (method and function are used interchangebly almost everywhere else).

What about `IRBody`?